plugins/module_utils/identity/bitwarden.py

```diff
@@ -0,0 +1,321 @@
+            raise AssertionError('Item not unique')
+
+        return matches[0]
+
+    def _dict_resolve_callback(self, value):
+        self._organization = Organization(id_=value['organizationId'])
+        self._folder = Folder(id_=value['folderId'])
+
+
+class Folder(_Element):
+    '''A folder in the password database.'''
+
+    organization = property(lambda self: self._organization)
+
+    def __init__(self, name=None, id_=None, organization=None, dict_=None):
+        if [name is None, id_ is None, dict_ is None].count(False) > 1:
+            raise AssertionError('Cannot specify folder in multiple ways')
+
+        self._name = name
+        self._id = id_
+        self._organization = organization
+
+        if dict_:
+            self.dict = dict_
+
+    def _by_attribute(self, attribute, value):
+        '''Return folder dict where the specified attribute
+        has the specified value.
+        '''
+        matches = [folder for folder in self.organization.folders if folder.dict[attribute] == value]
+
+        if not matches:
+            raise KeyError("Folder not found")
+        if len(matches) > 1:
+            raise AssertionError("Folder not unique")
+
+        return matches[0]
+
+    @property
+    def items(self):
+        query = ['list', 'items']
+
+        # Only restrict by folder if specified.
+        if self.name is not None:
+            # bw treats 'null' as items not belonging to any folder.
+            query.extend(['--folderid', self.id or 'null'])
+
+        # Filter by organization since `bw` ORs filters together.
+        return [Item(dict_=d) for d in Client.run_json(query) if self.organization.matches(d)]
+
+    def get_item(self, name=None, id_=None):
+        return Item(name, id_, self, self.organization)
+
+    def matches(self, folder):
+        '''Return whether folder matches a pattern indicated by self.
+
+        If self.name is None, any folder matches.
+        If self.name is '' or 'No Folder', only the null folder is matched.
+        All other folders are matched by id.
+        '''
+
+        if self.name is None:
+            return True
+
+        # No Folder is a special folder name in bw.
+        if self.name in ('', 'No Folder'):
+            return folder['folderId'] is None
+
+        return self.id == folder['folderId']
+
+
+class Organization(_Element):
+    '''An organization in the password database.'''
+
+    def __init__(self, name=None, id_=None, dict_=None):
+        if [name is None, id_ is None, dict_ is None].count(False) > 1:
+            raise AssertionError('Cannot specify organization in multiple ways')
+
+        self._name = name
+        self._id = id_
+
+        if dict_:
+            self.dict = dict_
+
+    def _by_attribute(self, attribute, value):
+        '''Return organization dict where the specified attribute
+        has the specified value.
+        '''
+
+        matches = [org for org in organizations() if org.dict[attribute] == value]
```

Comment:
Could you please elaborate on your reasoning?

IMHO, using `get` in this instance is less safe. If someone is searching for a non-existent attribute, `get` will return a value rather than generating an exception. On the other hand, I don't know enough about bitwarden to know if anyone would want to search for a Bitwarden attribute that isn't populated in record/entry.

If we're going to support querying over irregular entries, it would probably make more sense to include an explicit sentinel entry (for example, `org.dict.get(attribute, __anonymousObjctThatWillNeverMatchAnything)`). But then ... someone might want to treat non-populated values as `None`... then perhaps we'd want a setting somewhere to control non-existent-attribute-matching behaviour.

As such, I think the easiest thing to do is just scream at the API-user if they try to access an attribute that doesn't exist.

Reply:
`.get()` will return `None`, and `None` will always fail the `==` test. So, if an user asks for an organization that matches a non-existent attribute, they will get an empty list. I see no need to make that an error, but that's just me.

Reply:
Given what you've said here, I think we should leave it as an `[]` lookup instead of `get`. While the calls to the code here don't take advantage of it, the Bitwarden CLI does actually return JSON `null`s/Python `None`s -- including if one wants to specifically search for things that have no folder/organization. So to find, say, a folder with no organization, we can call `folder_obj._by_attribute('organizationid', None)`.

On a side note, in the Ansible module, we interpret `''` to be `None` and `None` to be a wildcard; to be a bit more consistent with the `bw` CLI, we could use something like `'*'` for a wildcard and `None` to mean no organization/folder, but I feel like not specifying an organization is a more intuitive way of saying "don't care" than "*". Any thoughts about that?

Reply:
> Given what you've said here, I think we should leave it as an `[]` lookup instead of `get`. While the calls to the code here don't take advantage of it, the Bitwarden CLI does actually return JSON `null`s/Python `None`s -- including if one wants to specifically search for things that have no folder/organization. So to find, say, a folder with no organization, we can call `folder_obj._by_attribute('organizationid', None)`.

Ok, that makes sense.

> On a side note, in the Ansible module, we interpret `''` to be `None` and `None` to be a wildcard; to be a bit more consistent with the `bw` CLI, we could use something like `'*'` for a wildcard and `None` to mean no organization/folder, but I feel like not specifying an organization is a more intuitive way of saying "don't care" than "*". Any thoughts about that?

While I am usually weary of these "translations", I understand that's often the case, and I don't have any better ideas for this particular situation. And yes, I agree that omitting the value is better than forcing the user to write a boilerplate wildcard value.